contributors/design-proposals/customresources-validation.md

```diff
@@ -0,0 +1,525 @@
+
+## Proposed Extension of CustomResourceDefinition
+
+We propose to add a field `validation` to the spec of a CustomResourceDefinition. As a first validation format we propose to use [JSON-Schema](http://json-schema.org/) under `CRD.Spec.Validation.JSONSchema`.
+
+JSON-Schema is a [standardized](https://tools.ietf.org/html/draft-zyp-json-schema-04) declarative specification language. Different keywords may be utilized to put constraints on the data. Thus it provides ways to make assertions about what a valid document must look like. 
+
+It is already used in Swagger/OpenAPI specs in Kubernetes and hence such a CRD specification integrates cleanly into the existing infrastructure of the API server which serves these specifications,
+* into kubectl which is able to verify YAML and JSON objects against the returned specification.
+* With the https://github.com/go-openapi/validate library, we have a powerful JSON-Schema validator which can be used client and server-side.
+
+## API Types
+
+The schema is referenced in [`CustomResourceDefinitionSpec`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L22). `Validation` is of the type `CustomResourceValidation`. The JSON-Schema is stored in a field of `Validation`. This way we can make the validation generic and add other validations in the future as well.
+
+The schema types follow those of the OpenAPI library, but we decided to define them independently for the API to have full control over the serialization and versioning. Hence, it is easy to convert our types into those used for validation or to integrate them into an OpenAPI spec later.
+
+Reference http://json-schema.org is also used by OpenAPI. We propose this as there are implementations available in Go and with OpenAPI, we will also be able to serve OpenAPI specs for CustomResourceDefintions.
+
+```go
+// CustomResourceSpec describes how a user wants their resource to appear
+type CustomResourceDefinitionSpec struct {
+    Group string `json:"group" protobuf:"bytes,1,opt,name=group"`
+    Version string `json:"version" protobuf:"bytes,2,opt,name=version"`
+    Names CustomResourceDefinitionNames `json:"names" protobuf:"bytes,3,opt,name=names"`
+    Scope ResourceScope `json:"scope" protobuf:"bytes,8,opt,name=scope,casttype=ResourceScope"`
+    // Validation describes the validation methods for CustomResources
+    Validation CustomResourceValidation `json:"validation,omitempty"`
+}
+
+// CustomResourceValidation is a list of validation methods for CustomResources
+type CustomResourceValidation struct {
+    // JSONSchema is the JSON Schema to be validated against.
+    // Can add other validation methods later if needed.
+    JSONSchema *JSONSchemaProps `json:"jsonSchema,omitempty"`
+}
+
+// JSONSchemaProps is a JSON-Schema following Specification Draft 4 (http://json-schema.org/).
+type JSONSchemaProps struct {
```

Comment:
I believe we don't really want to import here and end up making our types depend on external types.

Had considered using https://github.com/xeipuuv/gojsonschema but instead decided to go with go-openapi because we want to be able to express the schema via openapi docs.